### PR TITLE
[megatron] feat: LoRA adapter only refit (TensorLoRARequest)

### PR DESCRIPTION
**Pull request by HollowMan6**
### What does this PR do?

lora-performance
lora-critic-val-score
lora-actor-plus-rollout-mismatch

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/utils/config.py [hunk 1]
```diff
@@ -199,7 +199,10 @@ def check_mutually_exclusive(mbs, mbs_per_gpu, name: str):
         )
 
     # check LoRA rank in vLLM
-    if config.actor_rollout_ref.model.get("lora_rank", 0) > 0 and config.actor_rollout_ref.rollout.name == "vllm":
-        assert config.actor_rollout_ref.model.lora_rank <= 512, "LoRA rank in vLLM must be less than or equal to 512"
+    lora_rank = config.actor_rollout_ref.model.get("lora", {}).get("rank", 0)
+    if lora_rank <= 0:
+        lora_rank = config.actor_rollout_ref.model.get("lora_rank", 0)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/config.py]: We will eventually merge the splitted lora config

File: verl/workers/megatron_workers.py [hunk 2]
```diff
@@ -673,9 +673,24 @@ async def rollout_mode(self):
...
+        peft_config = None
+        do_lora_base_sync = False
+        if self.peft_cls is not None:
+            peft_config = build_peft_config_for_vllm(self.config.model.get("lora", {}))
+            # set sleep level for LoRA adapter weights only sync
+            # TODO: make this configurable so that users with small
+            # main memory can trade sync time to avoid OOM
+            self.rollout.sleep_level = 1
+
+            do_lora_base_sync = not self.base_sync_done or self.rollout.sleep_level != 1
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/megatron_workers.py:702]: This is for future configurations if users want to still set sleep level to 2 so that users with small main memory can trade sync time to avoid OOM

File: verl/workers/rollout/vllm_rollout/vllm_rollout.py [hunk 3]
```diff
@@ -122,11 +122,18 @@ def __init__(
...
+        if self.model_config.lora.get("merge", False):
+            lora_rank = 0
+        megatron_lora = True
+        if lora_rank <= 0:
+            megatron_lora = False
+            lora_rank = self.model_config.lora_rank
+        if lora_rank > 0:
+            self.lora_config = {"max_loras": 1, "max_lora_rank": get_vllm_max_lora_rank(lora_rank)}
+            if megatron_lora:
+                self.lora_config["fully_sharded_loras"] = True
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_rollout.py]: We will eventually merge the splitted lora config

**Comment by HollowMan6** [comment]
Now these PRs have all been merged into upstream, this PR is now ready for review:
- https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1766
- https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1788
- https://github.com/vllm-project/vllm/pull/31104

**Comment by HollowMan6** [comment]
Note that the changes in https://github.com/vllm-project/vllm/pull/31104 hasn't been landed into a release, so for now, we still set `merge` to True as the default value and maybe change it later

File: verl/utils/config.py [hunk 4]
```diff
@@ -199,7 +199,13 @@ def check_mutually_exclusive(mbs, mbs_per_gpu, name: str):
...
 
     # check LoRA rank in vLLM
-    if config.actor_rollout_ref.model.get("lora_rank", 0) > 0 and config.actor_rollout_ref.rollout.name == "vllm":
-        assert config.actor_rollout_ref.model.lora_rank <= 512, "LoRA rank in vLLM must be less than or equal to 512"
+    lora_config = config.actor_rollout_ref.model.get("lora", {})
+    lora_rank = lora_config.get("rank", 0)
+    if lora_config.get("merge", False):
+        lora_rank = 0
+    if lora_rank <= 0:
+        lora_rank = config.actor_rollout_ref.model.get("lora_rank", 0)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/config.py:207]: This is intended, as if we enable merge, we shouldn't enable LoRA on the vLLM side.

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py [hunk 5]
```diff
@@ -354,14 +351,22 @@ async def launch_server(self, master_address: str = None, master_port: int = Non
...
-                    "max_lora_rank": get_vllm_max_lora_rank(self.model_config.lora_rank),
-                }
-            )
+        lora_rank = self.model_config.lora.get("rank", 0)
+        megatron_lora = True
+        if self.model_config.lora.get("merge", False):
+            lora_rank = 0
+        if lora_rank <= 0:
+            megatron_lora = False
+            lora_rank = self.model_config.lora_rank
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_async_server.py:339]: This is intended, as if we enable merge, we shouldn't enable LoRA on the vLLM side.

File: verl/workers/rollout/vllm_rollout/vllm_rollout.py [hunk 6]
```diff
@@ -124,11 +124,18 @@ def __init__(
...
-            else {}
-        )
+        self.lora_config = {}
+        lora_rank = self.model_config.lora.get("rank", 0)
+        if self.model_config.lora.get("merge", False):
+            lora_rank = 0
+        megatron_lora = True
+        if lora_rank <= 0:
+            megatron_lora = False
+            lora_rank = self.model_config.lora_rank
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_rollout.py]: This is intended, as if we enable merge, we shouldn't enable LoRA on the vLLM side.

File: verl/utils/config.py [hunk 7]
```diff
@@ -199,7 +199,13 @@ def check_mutually_exclusive(mbs, mbs_per_gpu, name: str):
         )
 
     # check LoRA rank in vLLM
-    if config.actor_rollout_ref.model.get("lora_rank", 0) > 0 and config.actor_rollout_ref.rollout.name == "vllm":
-        assert config.actor_rollout_ref.model.lora_rank <= 512, "LoRA rank in vLLM must be less than or equal to 512"
+    lora_config = config.actor_rollout_ref.model.get("lora", {})
+    lora_rank = lora_config.get("rank", 0)
+    if lora_config.get("merge", False):
+        lora_rank = 0
+    if lora_rank <= 0:
```
Comment (ISEEKYAN) [inline verl/utils/config.py:206]: when would lora_rank be less than 0?

File: verl/utils/config.py [hunk 8]
```diff
@@ -199,7 +199,13 @@ def check_mutually_exclusive(mbs, mbs_per_gpu, name: str):
...
+        lora_rank = config.actor_rollout_ref.model.get("lora_rank", 0)
+    if lora_rank > 0 and config.actor_rollout_ref.rollout.name == "vllm":
+        assert lora_rank <= 512, "LoRA rank in vLLM must be less than or equal to 512"
```
Comment (ISEEKYAN) [inline verl/utils/config.py]: is it possible to use a number from vllm, instead of a hardcode 512? such as something like `assert lora_rank < vllm.xxx.xxx.max_lora_rank`

File: verl/utils/megatron_peft_utils.py [hunk 9]
```diff
@@ -15,9 +15,53 @@
 
 import os
 from pathlib import Path
+from typing import Iterator
 
 import torch
 
+# Map megatron lora target modules to HF-style module names for vLLM
```
Comment (ISEEKYAN) [inline verl/utils/megatron_peft_utils.py:22]: This is only an advice, not necessary, can we move the lora mapping to `megatron-bridge` instead of keeping them in `verl`?

File: verl/workers/megatron_workers.py [hunk 10]
```diff
@@ -345,6 +346,10 @@ def __init__(self, config: DictConfig, role: str, **kwargs):
         self._is_offload_grad = False
         self._is_offload_optimizer = False
 
+        # Initialize LoRA-related attributes (will be updated in _build_rollout if needed)
+        self.base_sync_done = False
```
Comment (ISEEKYAN) [inline verl/workers/megatron_workers.py:349]: is the `base_sync_done` an indicator of sync of only lora weights or the whole weights? I am not sure, this is ambiguous

File: verl/utils/config.py (hunk 7 above)
Comment (HollowMan6) [reply verl/utils/config.py:206]: There's no valid case for lora_rank to be less than 0. This is just an extra defense in case users want to pass values such as `-1` to currently megatron lora_rank in case they are using FSDP.

Comment (ISEEKYAN) [reply verl/utils/config.py:206]: get

File: verl/utils/config.py (hunk 8 above)
Comment (HollowMan6) [reply verl/utils/config.py]: I just checked the code and found this https://github.com/vllm-project/vllm/blob/841d53aaa8d674f2c9f72503e77f75e5ffa79c71/vllm/config/lora.py#L25, so yeah, I think we can do something as you mentioned

File: verl/utils/megatron_peft_utils.py (hunk 9 above)
Comment (HollowMan6) [reply verl/utils/megatron_peft_utils.py:22]: Yes, I think this is something in discussion, and I guess it would need some API designs.

File: verl/workers/megatron_workers.py (hunk 10 above)
Comment (HollowMan6) [reply verl/workers/megatron_workers.py:349]: This is a indicator only for the base weight. As currently, by default, vLLM will not load the base model weight but reply on weight sync before the training starts. Also, if sleep level 2 is turned on, the base weight will also need to get synced as they will get discarded when vLLM gets to sleep.

File: verl/utils/megatron_peft_utils.py (hunk 9 above)
Comment (ISEEKYAN) [reply verl/utils/megatron_peft_utils.py:22]: add @yaoyu-33 for vis. We need to extend bridge APIs with capability of exporting lora weights.

File: verl/utils/config.py (hunk 8 above)
Comment (HollowMan6) [reply verl/utils/config.py]: Updated via `get_vllm_max_lora_rank` for this!

**Comment by HollowMan6** [comment]
cc: @wuxibin89 @vermouth1992 for further comments
